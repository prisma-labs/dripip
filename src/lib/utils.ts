import { inspect } from 'util'

export function dump(arg: unknown): void {
  console.error(inspect(arg, { depth: 30 }))
}
/**
 * Indent a string that spans multiple lines.
 */
export function indentBlock(size: number, block: string): string {
  return block
    .split('\n')
    .map((line) => range(size).map(constant(' ')).join('') + line)
    .join('\n')
}

export const indentBlock4 = indentBlock.bind(null, 4)

/**
 * Create a function that will only ever return the given value when called.
 */
export const constant = <T>(x: T): (() => T) => {
  return function () {
    return x
  }
}

/**
 * Create a range of integers.
 */
const range = (times: number): number[] => {
  const list: number[] = []
  while (list.length < times) {
    list.push(list.length + 1)
  }
  return list
}

// type IndexableKeyTypes = string | number | symbol

// type Indexable<T = unknown> = Record<string | number, T>

// type JustIndexableTypes<T> = T extends IndexableKeyTypes ? T : never

// type KeysMatching<Rec, Keys> = NonNullable<
//   {
//     [RecKey in keyof Rec]: Rec[RecKey] extends Keys ? RecKey : never
//   }[keyof Rec]
// >

// export type GroupBy<T extends Indexable, K extends IndexableKeys<T>> = {
//   [KV in JustIndexableTypes<T[K]>]?: Array<T extends Record<K, KV> ? T : never>
// }

// type IndexableKeys<Rec> = KeysMatching<Rec, IndexableKeyTypes>

// export function groupByProp<
//   Obj extends Indexable,
//   KeyName extends IndexableKeys<Obj>
// >(xs: Obj[], keyName: KeyName): GroupBy<Obj, KeyName> {
//   type KeyValue = JustIndexableTypes<Obj[KeyName]>
//   const seed = {} as GroupBy<Obj, KeyName>

//   return xs.reduce((groupings, x) => {
//     const groupName = x[keyName] as KeyValue

//     if (groupings[groupName] === undefined) {
//       groupings[groupName] = []
//     }

//     // We know the group will exist, given above initializer.
//     groupings[groupName]!.push(
//       x as Obj extends Record<KeyName, KeyValue> ? Obj : never
//     )

//     return groupings
//   }, seed)
// }

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
export function casesHandled(x: never): never {
  throw new Error(`A case was not handled for value: ${x}`)
}

/**
 * Determine if the given array or object is empty.
 */
export function isEmpty(x: {} | unknown[]): boolean {
  return Array.isArray(x) ? x.length === 0 : Object.keys(x).length > 0
}

/**
 * Pause in time for given milliseconds.
 */
export function delay(milliseconds: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, 1000)
  })
}

/**
 * Like Array.findIndex but working backwards from end of array.
 */
export function findIndexFromEnd<T>(xs: T[], f: (x: T) => boolean): number {
  for (let index = xs.length - 1; index > -1; index--) {
    if (f(xs[index])) return index
  }
  return -1
}

/**
 * Get the last element of an array.
 */
export function last<T>(xs: T[]): null | T {
  if (xs.length === 0) return null
  return xs[xs.length - 1]
}

export function numericAscending(n1: number, n2: number): -1 | 0 | 1 {
  if (n1 < n2) return -1
  if (n1 > n2) return 1
  return 0
}
