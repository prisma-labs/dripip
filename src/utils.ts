import * as Git from 'simple-git/promise'
import * as Semver from 'semver'

// const emptyCommit = (git:Git.SimpleGit, options: string[]): Promise<Git.CommitSummary> => {
//   git.commit(["--allow-empty", "--message", "initial commit"])
// }

const getReleaseTagsAtCommit = (ref: string): Promise<Semver.SemVer[]> =>
  Git()
    .tag({ '--points-at': ref })
    .then(result =>
      result
        .trim()
        .split('\n')
        .map(tag => Semver.parse(tag))
        .filter((parsed): parsed is Semver.SemVer => parsed !== null)
    )

const indentBlock4 = (block: string): string => indentBlock(4, block)

const indentBlock = (size: number, block: string): string => {
  return block
    .split('\n')
    .map(
      line =>
        range(size)
          .map(constant(' '))
          .join('') + line
    )
    .join('\n')
}

const constant = <T>(x: T): (() => T) => {
  return function() {
    return x
  }
}

const range = (times: number): number[] => {
  const list: number[] = []
  while (list.length < times) {
    list.push(list.length + 1)
  }
  return list
}

export { getReleaseTagsAtCommit, indentBlock, indentBlock4 }
