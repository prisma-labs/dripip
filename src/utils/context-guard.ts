import { CLIError } from '@oclif/errors'
import * as Context from './context'

type Options = {
  context: Context.Context
}

/**
 *
 * @example
 *
 * ```ts
 * guard(ctx)
 *   .must(isTrunk())
 *   .check(flags.upToDate ? 'must' : 'prefer', upToDate())
 *
 * if (flags.npmPublish) {
 *   guard.must(publishRights())
 * }
 *
 * guard.run()
 * ```
 */
export function check(options: Options) {
  const contextualizedCheckers: Requirement[] = []

  const api = {
    check(level: Requirement['level'], validator: Validator) {
      contextualizedCheckers.push({ level, validator })
      return api
    },
    stopUnless(checker: Validator) {
      return api.check('prefer', checker)
    },
    errorUnless(checker: Validator) {
      return api.check('must', checker)
    },
    run() {
      return contextualizedCheckers.reduce((results, requirement) => {
        const resultSugar = requirement.validator.run(options.context)
        const result = typeof resultSugar === 'boolean' ? booleanResult(resultSugar) : resultSugar

        const group =
          result.kind === 'pass'
            ? results.passes
            : requirement.level === 'must'
            ? results.errors
            : results.stops

        group.push({
          details: result.details,
          summary: requirement.validator.summary,
          code: requirement.validator.code,
        })

        return results
      }, createReport())
    },
  }

  return api
}

/**
 *
 */
function createReport(): Report {
  return {
    stops: [],
    errors: [],
    passes: [],
  }
}

/**
 *
 */
interface EnforceInput {
  context: Context.Context
  report: Report
  json?: boolean
}

/**
 * Enforce the result of a report. Throws errors/stop-signal that will halt the
 * CLI if needed.
 */
export function guard(input: EnforceInput): void {
  if (input.report.errors.length) {
    if (input.json) {
      if (input.report.errors.length) {
        throw new JSONCLIError({
          context: input.context,
          failures: input.report.errors,
        })
      }
    } else {
      throw new PassthroughCLIError(
        input.report.errors
          .map((failure) => {
            return failure.summary
          })
          .join('\n - ')
      )
    }
  }

  if (input.report.stops.length) {
    throw new CLIStop({ reasons: input.report.stops, json: input.json })
  }
}

interface Requirement {
  validator: Validator
  level: 'prefer' | 'must'
}

export interface ValidationResult {
  code: Validator['code']
  summary: string
  details: Record<string, unknown>
}

interface Report {
  errors: ValidationResult[]
  stops: ValidationResult[]
  passes: ValidationResult[]
}

//
// Errors for halting CLI process
//

export class JSONCLIError extends CLIError {
  oclif!: { exit: number }
  code = 'JSONError'

  constructor(private errorObject: object) {
    super('...todo...', { exit: 1 })
  }

  render(): string {
    return JSON.stringify(this.errorObject)
  }
}

export class PassthroughCLIError extends CLIError {
  oclif!: { exit: number }
  code = 'PassthroughError'

  constructor(private errorString: string) {
    super('...todo...', { exit: 1 })
  }

  render(): string {
    return this.errorString
  }
}

export class CLIStop extends CLIError {
  oclif!: { exit: number }
  code = 'Stop'

  constructor(private input: { reasons: ValidationResult[]; json?: boolean }) {
    super('...todo...', { exit: 0 })
  }

  render(): string {
    if (this.input.json) {
      return JSON.stringify({ reasons: this.input.reasons })
    } else {
      return 'Nothing to do:\n\n' + this.input.reasons.map((r) => r.summary).join('\n -')
    }
  }
}

//
// Base Either-like & validator types
//

export interface Validator {
  code: string
  summary: string
  run(ctx: Context.Context): Pass | Fail | boolean
}

interface Pass {
  kind: 'pass'
  details: Record<string, unknown>
}

interface Fail {
  kind: 'fail'
  details: Record<string, unknown>
}

export type Result = Pass | Fail

export function booleanResult(bool: boolean): Result {
  return bool ? { kind: 'pass', details: {} } : { kind: 'fail', details: {} }
}
