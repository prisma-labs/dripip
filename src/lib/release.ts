import * as Semver from './semver'
import * as Git from './git'
import { findIndexFromEnd, last } from './utils'

/**
 * Get the previous stable and commits since then. If there is no previous
 * stable then commits are counted from start of history.
 */
export async function getCurrentSeries(git: Git.Simple): Promise<Series> {
  return getLog(git).then(buildSeries)
}

/**
 * Get the latest stable and all subsequent commits. If no stable then all
 * commits (edge-case, new project) since-inclusive initial commit. The returned
 * tuple contains the  stable commit partitioned from the subsequent commits.
 * The stable commit may be null.
 */
async function getLog(git: Git.Simple): Promise<SeriesLog> {
  const previousStableCommit = await findLatestStable(git)
  const commits = await Git.log(git, { since: previousStableCommit })
  return previousStableCommit
    ? [commits.shift() as Git.LogEntry, commits]
    : [null, commits]
}

export type Commit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null | Semver.StableVer
    preview: null | Semver.PreviewVer
  }
}

export type StableCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: Semver.StableVer
    preview: null
  }
}

export type PreviewCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null
    preview: Semver.PreviewVer
  }
}

export type MaybePreviewCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null
    preview: null | Semver.PreviewVer
  }
}

export type UnreleasedCommit = {
  message: string
  sha: string
  nonReleaseTags: string[]
  releases: {
    stable: null
    preview: null
  }
}

export type Series = {
  previousStable: null | StableCommit
  commitsSinceStable: MaybePreviewCommit[]
  previousPreview: null | PreviewCommit
  commitsSincePreview: UnreleasedCommit[]
  current: Commit
}

export type SeriesLog = [null | Git.LogEntry, Git.LogEntry[]]

/**
 * Build structured series data from a raw series log.
 */
export function buildSeries([previousStable, commitsSince]: SeriesLog): Series {
  if (previousStable === null && commitsSince.length === 0) {
    throw new Error(
      `Cannot build release series with given data. There is no previous stable release and no commits since. This would indicate an unexpected error or working with a git repo that has zero commits. The latter should be guarded by upstream checks. Therefore this is bad. There must be a bug.`
    )
  }

  const commitsSinceStable = commitsSince.map(c => {
    const { message, sha } = c
    return {
      message,
      sha,
      nonReleaseTags: c.tags.filter(isUnknownTag),
      releases: {
        stable: Semver.parse(c.tags.find(isStableTag) ?? ''),
        preview: Semver.parsePreview(c.tags.find(isPreviewTag) ?? ''),
      },
    } as MaybePreviewCommit
  })

  const prevPreviewI = findIndexFromEnd(
    commitsSinceStable,
    c => c.releases.preview !== null
  )

  let previousPreview: null | PreviewCommit = null
  let commitsSincePreview: UnreleasedCommit[] = []
  if (prevPreviewI !== -1) {
    previousPreview = commitsSinceStable[prevPreviewI]! as PreviewCommit
    commitsSincePreview = commitsSinceStable.slice(
      prevPreviewI + 1
    ) as UnreleasedCommit[]
  }

  const previousStable2 =
    previousStable === null
      ? null
      : ({
          sha: previousStable.sha,
          message: previousStable.message,
          nonReleaseTags: previousStable.tags.filter(isUnknownTag),
          releases: {
            stable: Semver.parse(previousStable.tags.find(isStableTag)!)!,
            preview: Semver.parsePreview(
              previousStable.tags.find(isPreviewTag) ?? ''
            ),
          },
        } as StableCommit)

  return {
    previousStable: previousStable2,
    commitsSinceStable,
    previousPreview,
    commitsSincePreview,
    // If there are no commits since stable and no stable that means we're on a
    // repo with no commit. This edge-case is ignored. It is assumed that it
    // will be validated for before calling this function.
    current: last(commitsSinceStable) ?? previousStable2!,
  }
}

function isUnknownTag(tag: string): boolean {
  return Semver.parse(tag) === null
}

function isStableTag(tag: string): boolean {
  const v = Semver.parse(tag)
  if (v === null) return false
  return Semver.isStable(v)
}

function isPreviewTag(tag: string): boolean {
  const v = Semver.parse(tag)
  if (v === null) return false
  return Semver.isPreview(v)
}

//
// Private helpers
//

async function findLatestStable(git: Git.Simple): Promise<null | string> {
  return Git.findTag(git, { matcher: isStableTag })
}

// const zeroReleases = {
//   stable: null,
//   preview: null,
// }

/**
 * Get the releases at the given commit.
 */
// export async function getReleasesAtCommit(
//   sha: string
// ): Promise<CommitReleases> {
//   const git = createGit()
//   const tags = await Git.gitGetTags(git, { ref: sha })
//   if (isEmpty(tags)) return zeroReleases
//   const parsedTags = groupByProp(tags.map(parseTag), 'type')
//   const stableTags = parsedTags.stable_release ?? []
//   const previewtags = parsedTags.pre_release ?? []

//   const invariantViolations = []
//   if (stableTags.length > 1) {
//     invariantViolations.push(
//       `Multiple stable releases found but there should only be 0 or 1: ${stableTags}`
//     )
//   }
//   if (previewtags.length > 1) {
//     invariantViolations.push(
//       `Multiple preview releases found but there should only be 0 or 1: ${stableTags}`
//     )
//   }
//   if (invariantViolations.length > 0) {
//     throw new Error(
//       `While getting the pre-existing releases at commit ${sha} the following invariant violations were found:\n\n${invariantViolations
//         .map(x => `    - ${x}`)
//         .join('\n')}`
//     )
//   }

//   return {
//     stable: stableTags[0]
//       ? {
//           type: 'stable',
//           version: stableTags[0].value,
//           sha,
//         }
//       : null,
//     preview: previewtags[0]
//       ? {
//           type: 'preview',
//           version: previewtags[0].value,
//           sha,
//           buildNum: previewtags[0].value.prerelease[1] as number,
//         }
//       : null,
//   }
// }

// export type ParsedTag =
//   | { type: 'unknown'; value: string }
//   | { type: 'stable_release'; value: SemVer.SemVer }
//   | { type: 'pre_release'; value: SemVer.SemVer }

// export function parseTag(rawTag: string): ParsedTag {
//   const semverParseResult = SemVer.parse(rawTag)
//   if (semverParseResult !== null) {
//     if (semverParseResult.prerelease.length > 0) {
//       return { type: 'pre_release', value: semverParseResult } as const
//     } else {
//       return { type: 'stable_release', value: semverParseResult } as const
//     }
//   } else {
//     return { type: 'unknown', value: rawTag } as const
//   }
// }
