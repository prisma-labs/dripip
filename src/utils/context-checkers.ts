import { validateNPMAuthSetup } from '../lib/npm-auth'
import { Validator } from './context-guard'

export function isTrunk(): Validator {
  return {
    code: 'must_be_on_trunk',
    summary: 'You must be on the trunk branch',
    run(ctx) {
      return ctx.currentBranch.isTrunk
    },
  }
}

export function branchSynced(): Validator {
  return {
    code: 'branch_not_synced_with_remote',
    summary: 'Your branch must be synced with the remote',
    run(ctx) {
      if (ctx.currentBranch.syncStatus === 'synced') {
        return true
      } else {
        return {
          kind: 'fail',
          details: {
            syncStatus: ctx.currentBranch.syncStatus,
          },
        }
      }
    },
  }
}

export function npmAuthSetup(): Validator {
  return {
    code: 'npm_auth_not_setup',
    summary: 'You must have npm auth setup to publish to the registry',
    run() {
      const result = validateNPMAuthSetup()

      if (result.kind === 'pass') {
        return true
      } else {
        return { kind: 'fail', details: { reasons: result.reasons } }
      }
    },
  }
}
