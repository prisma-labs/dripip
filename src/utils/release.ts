import * as ConventionalCommit from '../lib/conventional-commit'
import * as Git from '../lib/git'
import * as Semver from '../lib/semver'

export type Release = {
  bumpType: Semver.MajMinPat
  version: Semver.Ver
}

export function shortSha(c: Commit): string {
  return c.sha.slice(0, 7)
}

/**
 * Get the previous stable and commits since then. If there is no previous
 * stable then commits are counted from start of history.
 */
export async function getCurrentSeries(input: { cwd: string }): Promise<Series> {
  return getLog({ cwd: input.cwd }).then(buildSeries)
}

/**
 * Like `getLog` but works on the given data. Useful for unit testing with mock
 * log data.
 */
export function fromLogs(entries: Git.LogEntry[]): Series {
  const commits: Git.LogEntry[] = []
  let previousStableCommit: null | Git.LogEntry = null

  for (const log of entries) {
    if (log.tags.find(isStableTag)) {
      previousStableCommit = log
      break
    }
    commits.push(log)
  }

  return buildSeries([previousStableCommit, commits])
}

/**
 * Get the latest stable and all subsequent commits. If no stable then all
 * commits (edge-case, new project) since-inclusive initial commit. The returned
 * tuple contains the  stable commit partitioned from the subsequent commits.
 * The stable commit may be null.
 */
async function getLog(input: { cwd: string }): Promise<SeriesLog> {
  const commits: Git.LogEntry[] = []
  let previousStableCommit: null | Git.LogEntry = null

  for await (const log of Git.streamLog({ cwd: input.cwd })) {
    if (log.tags.find(isStableTag)) {
      previousStableCommit = log
      break
    }
    commits.push(log)
  }

  return [previousStableCommit, commits]
}

/**
 * Get the current commit.
 */
export async function getCurrentCommit(): Promise<Commit> {
  let currLog: undefined | Git.LogEntry
  for await (const log of Git.streamLog()) {
    currLog = log
    break
  }

  if (!currLog) {
    throw new Error('There are no commits in this repo')
  }

  return logEntryToCommit(currLog)
}

// todo non-conforming aka. non-conventional commits
type CommitBase = {
  message: {
    raw: string
    parsed: null | ConventionalCommit.ConventionalCommit
  }
  sha: string
  nonReleaseTags: string[]
}

export type Commit = CommitBase & {
  releases: {
    stable: null | Semver.StableVer
    preview: null | Semver.PreviewVer
  }
}

export type StableCommit = CommitBase & {
  releases: {
    stable: Semver.StableVer
    preview: null
  }
}

export type PreviewCommit = CommitBase & {
  releases: {
    stable: null
    preview: Semver.PreviewVer
  }
}

export type MaybePreviewCommit = CommitBase & {
  releases: {
    stable: null
    preview: null | Semver.PreviewVer
  }
}

export type UnreleasedCommit = CommitBase & {
  releases: {
    stable: null
    preview: null
  }
}

export type Series = {
  previousStable: null | StableCommit
  commitsInNextStable: MaybePreviewCommit[]
  previousPreview: null | PreviewCommit
  commitsInNextPreview: UnreleasedCommit[]
  current: Commit
  hasBreakingChange: boolean
  isInitialDevelopment: boolean
}

export type SeriesLog = [null | Git.LogEntry, Git.LogEntry[]]

/**
 * Transform a log entry into a commit.
 */
function logEntryToCommit(log: Git.LogEntry): Commit {
  const parsedMessage = ConventionalCommit.parse(log.message)
  const previewTag = log.tags.find(isPreviewTag) ?? null
  const stableTag = log.tags.find(isStableTag) ?? null
  const otherTags = log.tags.filter(isUnknownTag)

  return {
    sha: log.sha,
    message: {
      parsed: parsedMessage,
      raw: log.message,
    },
    releases: {
      stable: stableTag ? Semver.parse(stableTag) : null,
      preview: previewTag ? Semver.parsePreview(previewTag) : null,
    },
    nonReleaseTags: otherTags,
  }
}

/**
 * Build structured series data from a raw series log.
 */
export function buildSeries([previousStableLogEntry, commitsSincePrevStable]: SeriesLog): Series {
  if (previousStableLogEntry === null && commitsSincePrevStable.length === 0) {
    throw new Error(
      `Cannot build release series with given data. There is no previous stable release and no commits since. This would indicate an unexpected error or working with a git repo that has zero commits. The latter should be guarded by upstream checks. Therefore this is bad. There must be a bug.`
    )
  }

  let hasBreakingChange = false

  const commitsInNextStable = commitsSincePrevStable.map((c) => {
    const parsedMessage = ConventionalCommit.parse(c.message)
    if (parsedMessage?.breakingChange) {
      hasBreakingChange = true
    }
    return {
      message: {
        raw: c.message,
        parsed: parsedMessage,
      },
      sha: c.sha,
      nonReleaseTags: c.tags.filter(isUnknownTag),
      releases: {
        stable: Semver.parse(c.tags.find(isStableTag) ?? ''),
        preview: Semver.parsePreview(c.tags.find(isPreviewTag) ?? ''),
      },
    } as MaybePreviewCommit
  })

  const previousPreviewIndex = commitsInNextStable.findIndex((c) => c.releases.preview !== null)

  const previousPreview =
    previousPreviewIndex === -1 ? null : (commitsInNextStable[previousPreviewIndex]! as PreviewCommit)

  const commitsInNextPreview: UnreleasedCommit[] =
    previousPreviewIndex === -1
      ? (commitsInNextStable as UnreleasedCommit[])
      : (commitsInNextStable.slice(0, previousPreviewIndex) as UnreleasedCommit[])

  const previousStable =
    previousStableLogEntry === null
      ? null
      : ({
          sha: previousStableLogEntry.sha,
          message: {
            raw: previousStableLogEntry.message,
            parsed: ConventionalCommit.parse(previousStableLogEntry.message),
          },
          nonReleaseTags: previousStableLogEntry.tags.filter(isUnknownTag),
          releases: {
            stable: Semver.parse(previousStableLogEntry.tags.find(isStableTag)!)!,
            preview: Semver.parsePreview(previousStableLogEntry.tags.find(isPreviewTag) ?? ''),
          },
        } as StableCommit)

  const isInitialDevelopment = (previousStable?.releases.stable.major ?? 0) < 1

  return {
    isInitialDevelopment,
    hasBreakingChange,
    previousStable: previousStable,
    previousPreview,
    commitsInNextStable,
    commitsInNextPreview,
    // If there are no commits since stable and no stable that means we're on a
    // repo with no commit. This edge-case is ignored. It is assumed that it
    // will be validated for before calling this function.
    current: commitsInNextStable[0] ?? previousStable!,
  }
}

function isUnknownTag(tag: string): boolean {
  return Semver.parse(tag) === null
}

function isStableTag(tag: string): boolean {
  const v = Semver.parse(tag)
  if (v === null) return false
  return Semver.isStable(v)
}

function isPreviewTag(tag: string): boolean {
  const v = Semver.parse(tag)
  if (v === null) return false
  return Semver.isPreview(v)
}

// function getNextVer(series: Series): Semver.PreviewVer {
//   if (series.previousPreview === null) {
//     return series.previousPreview
//   }
// }

export type NoReleaseReason = 'empty_series' | 'no_meaningful_change'

export function isNoReleaseReason(x: unknown): x is NoReleaseReason {
  return x === 'no_meaningful_change' || x === 'empty_series'
}

/**
 * Get the next stable from a given series. If a release cannot be made for the
 * given series then a code indicating why will be returned.
 */
export function getNextStable(series: Series): NoReleaseReason | Release {
  if (series.commitsInNextStable.length === 0) {
    return 'empty_series'
  }

  const bumpType = ConventionalCommit.calcBumpType(
    series.isInitialDevelopment,
    series.commitsInNextStable.map((c) => c.message.raw)
  )

  if (bumpType === null) return 'no_meaningful_change'

  const version = Semver.incStable(
    bumpType,
    series.previousStable === null ? Semver.zeroVer : series.previousStable.releases.stable
  )

  return { bumpType, version }
}

/**
 * Calculate the preview release for the given series.
 */
export function getNextPreview(series: Series): NoReleaseReason | Release {
  if (series.commitsInNextStable.length === 0) {
    return 'empty_series'
  }

  // todo test this case, it was fixed without regression test being added
  const bumpTypeContributionFromCommitsInNextPreview = ConventionalCommit.calcBumpType(
    series.isInitialDevelopment,
    series.commitsInNextPreview.map((c) => c.message.raw)
  )

  if (bumpTypeContributionFromCommitsInNextPreview === null) return 'no_meaningful_change'

  const bumpType = ConventionalCommit.calcBumpType(
    series.isInitialDevelopment,
    series.commitsInNextStable.map((c) => c.message.raw)
  )

  // Stable is superset of preview so based on guard above there MUST be a
  // bumpType here.
  if (!bumpType) {
    throw new Error(
      'next preview in series has bump type but not series as a whole. You should not be seeing this. This should be impossible.'
    )
  }

  const nextStable = Semver.incStable(bumpType, series.previousStable?.releases.stable ?? Semver.zeroVer)

  const version = Semver.stableToPreview(
    nextStable,
    'next',
    (series.previousPreview?.releases.preview.preRelease.buildNum ?? Semver.zeroBuildNum) + 1
  )

  return { bumpType, version }
}

//
// Private helpers
//

// async function findLatestStable(git: Git.Simple): Promise<null | string> {
//   return Git.findTag(git, { matcher: isStableTag })
// }

// const zeroReleases = {
//   stable: null,
//   preview: null,
// }

/**
 * Get the releases at the given commit.
 */
// export async function getReleasesAtCommit(
//   sha: string
// ): Promise<CommitReleases> {
//   const git = createGit()
//   const tags = await Git.gitGetTags(git, { ref: sha })
//   if (isEmpty(tags)) return zeroReleases
//   const parsedTags = groupByProp(tags.map(parseTag), 'type')
//   const stableTags = parsedTags.stable_release ?? []
//   const previewtags = parsedTags.pre_release ?? []

//   const invariantViolations = []
//   if (stableTags.length > 1) {
//     invariantViolations.push(
//       `Multiple stable releases found but there should only be 0 or 1: ${stableTags}`
//     )
//   }
//   if (previewtags.length > 1) {
//     invariantViolations.push(
//       `Multiple preview releases found but there should only be 0 or 1: ${stableTags}`
//     )
//   }
//   if (invariantViolations.length > 0) {
//     throw new Error(
//       `While getting the pre-existing releases at commit ${sha} the following invariant violations were found:\n\n${invariantViolations
//         .map(x => `    - ${x}`)
//         .join('\n')}`
//     )
//   }

//   return {
//     stable: stableTags[0]
//       ? {
//           type: 'stable',
//           version: stableTags[0].value,
//           sha,
//         }
//       : null,
//     preview: previewtags[0]
//       ? {
//           type: 'preview',
//           version: previewtags[0].value,
//           sha,
//           buildNum: previewtags[0].value.prerelease[1] as number,
//         }
//       : null,
//   }
// }

// export type ParsedTag =
//   | { type: 'unknown'; value: string }
//   | { type: 'stable_release'; value: SemVer.SemVer }
//   | { type: 'pre_release'; value: SemVer.SemVer }

// export function parseTag(rawTag: string): ParsedTag {
//   const semverParseResult = SemVer.parse(rawTag)
//   if (semverParseResult !== null) {
//     if (semverParseResult.prerelease.length > 0) {
//       return { type: 'pre_release', value: semverParseResult } as const
//     } else {
//       return { type: 'stable_release', value: semverParseResult } as const
//     }
//   } else {
//     return { type: 'unknown', value: rawTag } as const
//   }
// }
