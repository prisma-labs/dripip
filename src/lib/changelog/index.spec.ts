import stripAnsi from 'strip-ansi'
import * as Release from '../../utils/release'
import * as Changelog from './'

const mockCommits: Release.MockCommit[] = [
  { message: 'chore: b' },
  { message: 'fix: b' },
  { message: 'feat: b' },
  { message: 'refactor: a' },
  { message: 'non conforming commit b' },
  { message: 'chore: a' },
  { message: 'feat: a\n\nBREAKING CHANGE:\ntoto' },
  { message: 'perf: a' },
  { message: 'fix: a' },
  { message: 'non conforming commit a' },
  { message: 'feat: blah', version: '0.1.0' },
]

describe('.render() with markdown', () => {
  const render = (...commits: Release.MockCommit[]) => {
    // return Changelog.Renderers.Markdown.render(Changelog.fromSeries(Release.fromMockCommits(commits)))
    return Changelog.render(Changelog.fromSeries(Release.fromMockCommits(commits)), { as: 'markdown' })
  }
  it('renders release changelog for the current release series', () => {
    const changelog = render(...mockCommits)
    expect(changelog).toMatchSnapshot()
  })
  it('shows breaking commits twice, once in breaking section and once in its native section', () => {
    const changelog = render({ message: 'a: foo\n\nBREAKING CHANGE:\nbar' })
    expect(changelog).toMatch(/BREAKING CHANGES\n\n.*foo/)
    expect(changelog).toMatch(/Improvements\n\n.*foo/)
  })
  it('shows breaking label if commit breaking except within breaking section', () => {
    const changelog = render({ message: 'a: foo\n\nBREAKING CHANGE:\nbar' })
    expect(changelog).toMatch(/Improvements\n\n.*(breaking).*foo/)
  })
  it('hides unknown section if no unknown commits', () => {
    const changelog = render({ message: 'a: b' })
    expect(changelog).not.toMatch(Changelog.empty().unspecified.label)
  })
  it('improvements section prefixes commits with their type', () => {
    const changelog = render({ message: 'a: foo' })
    expect(changelog).toMatch(/a: foo/)
  })
})

describe('.render() with terminal', () => {
  const render = (...commits: Release.MockCommit[]) => {
    return Changelog.render(Changelog.fromSeries(Release.fromMockCommits(commits)), { as: 'plain' })
  }

  it('renders changelog for the current release series', () => {
    const changelog = stripAnsi(render(...mockCommits))
    expect(changelog).toMatchSnapshot()
  })
})
