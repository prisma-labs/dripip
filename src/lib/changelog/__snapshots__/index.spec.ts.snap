// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.render() with markdown renders release changelog for the current release series 1`] = `
"#### BREAKING CHANGES

- shasha# a

#### Features

- shasha# b
- shasha# (breaking) a

#### Fixes

- shasha# b
- shasha# a

#### Improvements

- shasha# refactor: a
- shasha# perf: a

#### Chores

- shasha# b
- shasha# a

#### Unspecified Changes

- shasha# non conforming commit b
- shasha# non conforming commit a
"
`;

exports[`.render() with terminal renders changelog for the current release series 1`] = `
"BREAKING CHANGES

  shasha# a

Features

  shasha# b
  shasha# (breaking) a

Fixes

  shasha# b
  shasha# a

Improvements

  shasha# refactor: a
  shasha# perf: a

Chores

  shasha# b
  shasha# a

Unspecified Changes

  shasha# non conforming commit b
  shasha# non conforming commit a
"
`;
